src/libostree/ostree-sysroot-deploy.c

```diff
@@ -973,17 +978,88 @@ checksum_from_kernel_src (const char   *name,
 }
 
 static gboolean
-syncfs_dir_at (int            dfd,
-               const char    *path,
-               GCancellable  *cancellable,
-               GError       **error)
+fsfreeze_thaw_cycle (int            rootfs_dfd,
+                     GCancellable *cancellable,
+                     GError       **error)
 {
-  glnx_fd_close int child_dfd = -1;
-  if (!glnx_opendirat (dfd, path, TRUE, &child_dfd, error))
+  GLNX_AUTO_PREFIX_ERROR ("During fsfreeze-thaw", error);
+
+  int pipefds[2];
+  if (!g_unix_open_pipe (pipefds, FD_CLOEXEC, error))
     return FALSE;
-  if (syncfs (child_dfd) != 0)
-    return glnx_throw_errno_prefix (error, "syncfs(%s)", path);
+  glnx_fd_close int pipe_read = pipefds[0];
+  glnx_fd_close int pipe_write = pipefds[1];
+
+  pid_t pid = fork ();
+  if (pid < 0)
+    return glnx_throw_errno_prefix (error, "fork");
 
+  char c = '!';
+  if (pid == 0) /* Child watchdog/unfreezer process. */
+    {
+      if (daemon (0, 0) < 0)
+        err (1, "daemon");
+      int sigs[] = { SIGINT, SIGTERM };
+      for (guint i = 0; i < G_N_ELEMENTS (sigs); i++)
+        {
+          if (signal (sigs[i], SIG_IGN) == SIG_ERR)
+            err (1, "signal");
+        }
+      if (TEMP_FAILURE_RETRY (write (pipe_write, &c, sizeof (c))) < 0)
+        err (1, "write");
+      /* Wait for the parent to say it's going to freeze. */
+      (void) TEMP_FAILURE_RETRY (read (pipefds[0], &c, sizeof (c)));
```

Comment:
Wait, this is referring to the same pipe, though right? Won't we just pick up the data we just wrote up above? If we want bidirectional comms, I think we need two pipes, right? (So 4 fds total).